Call for Testing: Speeding up compilation with `hint-mostly-unused` (#1662)
rust-lang/blog.rust-lang.org: content/inside-rust/call-for-testing-hint-mostly-unused.md

top-level crate.

In some cases, crates add feature flags to control compilation of their API
surface. This can improve compile time, but adds complexity for users, who now
need to determine which features they need for the APIs they use. Features also
constitute a stable interface of a crate, and changing feature flags can be a
breaking change. And even with feature flags, not every enabled function will
be needed; there is a balance between granularity and ease of use.

## Deferring code generation with `-Zhint-mostly-unused`

The latest nightly `rustc` compiler now supports an option
`-Zhint-mostly-unused`, which tells `rustc` that the crate's API surface will
mostly go unused. This is a hint, and `rustc` doesn't make guarantees about its
exact behavior (so that we can extend or improve it in the future), but
currently it causes the compiler to defer as much of code generation as
possible. (The compiler does this using the same machinery as
`#[inline(always)]`.)

Applying this option to key crates you depend on (and use only a small subset
of) can provide a substantial reduction in compile time, for debug builds and
especially for release builds.

## How does this perform?

> | **Dependency Crate** | **Before** | **`hint-mostly-unused`** | **Delta** |

Author: I'll add notes that it's a clean release build.

I don't expect this to *hurt* incremental builds, in practice. It might depend on how much of the crate's surface area you actually use. But I expect that real-world performance numbers would help there. I'll add some notes on that.

+++
path = "inside-rust/2025/07/14/call-for-testing-hint-mostly-unused"
title = "Call for Testing: Speeding up compilation with `hint-mostly-unused`"
authors = ["Josh Triplett"]
+++

I'm pleased to announce, and call for testing of, the nightly-only `rustc`
`-Zhint-mostly-unused` option, and the corresponding nightly Cargo features
`profile.hint-mostly-unused` and `hints.mostly-unused`. These options can help
accelerate your Rust compile time in some cases, by avoiding compilation of
items from your dependencies that you aren't using. Your feedback will help
evaluate these features and make progress towards stabilizing them in the
future.

## Background

> Some crates provide comprehensive APIs with a very large surface area, yet many

Reviewer: Perhaps it will help intuition to say this is for APIs that are monomorphic? Not sure how to put that.

Author: I attempted to cover this; let me know how the new version looks.

currently it causes the compiler to defer as much of code generation as
possible.

Applying this option to key crates you depend on (and use only a small subset
of) can provide a substantial reduction in compile time, for debug builds and
especially for release builds.

## How does this perform?

Some build timings for clean release builds of a crate depending on various
specific large API crates:

| **Dependency Crate** | **Before** | **`hint-mostly-unused`** | **Delta** |
| :- | -: | -: | -: |
| `windows`, all Graphics/UI features | 18.3s | 10.7s | -42% |
| `windows`, all features | 3m 48s | 2m 55s | -23% |
| `rustix`, `all-apis` feature | 5.9s | 4.3s | -27% |
| `x11rb` and `x11rb-protocol` | 5.3s | 2.6s | -51% |
| `aws-sdk-ec2` | 4m 07s | 2m 04s | -50% |

This performance improvement comes from deferring code generation. For
instance, the `windows` crate in the first example goes from building in 15.1s
of which 49% is codegen, to building in 7.5s of which 1% is codegen.

Note that this option does not provide a universal performance improvement for
every crate. Using it for crates whose API surface is mostly used, and/or used

> in multiple different crates or binaries (e.g. multiple test binaries that each

Author: This paragraph does also explicitly say it's a bad idea for crates whose API surface is mostly used.

I'll try splitting the paragraph up a bit, to better emphasize both conditions independently rather than as an "and/or".

+++
path = "inside-rust/2025/07/14/call-for-testing-hint-mostly-unused"
title = "Call for Testing: Speeding up compilation with `hint-mostly-unused`"
authors = ["Josh Triplett"]
+++

I'm pleased to announce, and call for testing of, the nightly-only `rustc`
`-Zhint-mostly-unused` option, and the corresponding nightly Cargo features
`profile.hint-mostly-unused` and `hints.mostly-unused`. These options can help

> accelerate your Rust compile time in some cases, by avoiding compilation of

Reviewer: "by avoiding compilation" may be interpreted as saying we will skip type-checking and/or borrow-checking. Which is not true in the current implementation. But maybe we want to leave open the possibility of doing so in the future?

I don't have a particular suggestion or concern here, I just want to make sure you mean the implication I'm reading.

Author: I definitely didn't intend that interpretation, no. I've rephrased accordingly.

items from your dependencies that you aren't using. Your feedback will help
evaluate these features and make progress towards stabilizing them in the
future.

## Background

When building a Rust library crate, the compiler generates compiled code for as

> much of the crate as it can (everything that isn't generic and isn't inlined),

Reviewer 2: ```suggestion
much of the crate as it can (everything that isn't generic and doesn't have `#[inline]`),
```

We should avoid using the adjective "inlined" for functions that have the inline attribute, as that adjective is more commonly used for functions that *actually have been inlined by the optimizer*. Let's avoid this confusion and be very explicit about the attribute.

which gets linked into later crates into the dependency graph. However, some
crates provide comprehensive APIs with a very large surface area, yet many of
their users need only a few entry points. In such cases, the compiler currently
spends time generating code for the entire crate, and the linker then ends up
throwing most of that code away as unused.

This can waste a substantial amount of compile time. Some large crates can take
minutes to compile, and when you use these large crates as dependencies, they

> can take a disproportionate amount of the entire compilation time of your
> top-level crate.

Reviewer: ```suggestion
can take a disproportionate fraction of the entire compilation time for a full build.
```
Because adding `-Zhint-mostly-unused` will increase the compile time of the top-level crate, while lowering the compile time of the entire multi-crate build.

possible.

Applying this option to key crates you depend on (and use only a small subset
of) can provide a substantial reduction in compile time, for debug builds and
especially for release builds.

## How does this perform?

Some build timings for clean release builds of a crate depending on various
specific large API crates:

| **Dependency Crate** | **Before** | **`hint-mostly-unused`** | **Delta** |
| :- | -: | -: | -: |
| `windows`, all Graphics/UI features | 18.3s | 10.7s | -42% |
| `windows`, all features | 3m 48s | 2m 55s | -23% |
| `rustix`, `all-apis` feature | 5.9s | 4.3s | -27% |
| `x11rb` and `x11rb-protocol` | 5.3s | 2.6s | -51% |
| `aws-sdk-ec2` | 4m 07s | 2m 04s | -50% |

This performance improvement comes from deferring code generation. For
instance, the `windows` crate in the first example goes from building in 15.1s
of which 49% is codegen, to building in 7.5s of which 1% is codegen.

Note that this option does not provide a universal performance improvement for
every crate; if used when not applicable, this option can make builds much
*slower*. Deferring compilation of the items in a crate can lead to redoing

> code generation for those items repeatedly. In particular, avoid using this
> hint for crates whose API surface is mostly used, and/or used in multiple

Reviewer: ```suggestion
code generation for those items repeatedly. In particular, this hint will
probably regress compile time if applied to crates whose API surface is
mostly used, and/or used in multiple
```

different crates or binaries (e.g. multiple test binaries that each test a
substantial swath of the API).

Always do performance analysis when considering this hint, and only apply it if
it applies obvious and substantial wins for your users. Never apply it across
the board to all your dependencies.

If most of the items in your crate are polymorphic (generic), this hint may be
redundant, as Rust already defers compilation of polymorphic items until they
get monomorphized with specific types.

> Also note that this only provides a performance win if you are rebuilding the

Reviewer 3: Is this true that it only helps when you are _rebuilding_ the dependency, or does it help whenever you are _building_ the dependency (ie even the first time)?
